Flaky test fix: CorfuReplicationClusterConfigIT.testBackupRestoreWorkflow

xcchang opened the pull request:
## Overview

Description:

Flaky test fix:
* `CorfuReplicationClusterConfigIT.testBackupRestoreWorkflow()`: Cannot reproduce locally with hundreds of reruns. Found a potential cause while examining the code in `DefaultClusterManager`. When we bring up backup cluster, we just put a sleep interval for the topology config to update, which could make it possible that for following forced snapshot sync is still performed by original source (state config cached in DefaultClusterManager) (#3468)

Why should this be merged: 

Related issue(s) (if applicable): #3468


## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

souravgarg commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -1630,9 +1630,6 @@ public void testBackupRestoreWorkflow() throws Exception {
         updateTopology(activeCorfuStore, DefaultClusterManager.OP_BACKUP);
         log.info("Change the topology!!!");
 
-        TimeUnit.SECONDS.sleep(shortInterval);
```
Yeah depending on sleep would make the code race prone. I'm wondering about the motivation behind this sleep.

xcchang replied:
I thought when this sleep was introduced, the purpose was to give the new topology time to take effect (writing to test purpose config table -> DefaultClusterManager update topology -> topology change event discovered by CorfuReplicationDiscoveryService).

From local running I never hit the flakiness of this test, but we did observe several failures in GitHub Action. I guess this could be a possible reason as if DefaultClusterManager hasn't updated topology, and the new forced snapshot sync has been requested, LR will be performed by the original source cluster instead of the backup cluster

souravgarg commented on test/src/test/java/org/corfudb/infrastructure/logreplication/LogReplicationFSMTest.java:
```diff
@@ -253,7 +253,9 @@ private void testSnapshotSender(int batchSize) throws Exception {
-        UUID snapshotSyncRequestId = transition(LogReplicationEventType.SNAPSHOT_SYNC_REQUEST, LogReplicationStateType.IN_SNAPSHOT_SYNC);
+        LogReplicationEvent event = new LogReplicationEvent(LogReplicationEventType.SNAPSHOT_SYNC_REQUEST);
+        UUID snapshotSyncRequestId = event.getEventId();
+        fsm.input(event);
```
Does fsm.input would take care of the transition?

xcchang replied:
Yes in this test class we emulate the state transition (by inputting events directly to fsm). Here the check is for when we input a SNAPSHOT_SYNC_REQUEST, verify the snapshot sync actually happens (we wait for WAIT_SNAPSHOT_APPLY state below).

Before this change we were also checking IN_SNAPSHOT_SYNC state, but that state could be already transited into WAIT_SNAPSHOT_APPLY state at the check point (which is automatically handled by fsm)

souravgarg commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/transport/sample/GRPCLogReplicationServerChannelAdapter.java:
```diff
@@ -36,7 +37,8 @@ public GRPCLogReplicationServerChannelAdapter(ServerContext serverContext, LogRe
         this.port = Integer.parseInt((String) serverContext.getServerConfig().get("<port>"));
-        this.server = ServerBuilder.forPort(port).addService(service).build();
+        this.server = ServerBuilder.forPort(port).addService(service)
+                .executor(Executors.newSingleThreadScheduledExecutor()).build();
```
Trivial question. Just for my knowledge. Can you please describe the reason on why do we need single threaded server?

xcchang replied:
https://github.com/CorfuDB/CorfuDB/blob/003d656cfe17f35a3d6814bd9b31e1c16e679b6b/infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/SnapshotSender.java#L144-L157
Sender side expects ACKs to be received in order, otherwise exception will be thrown. Actually in our netty implementation for rpc handlers, the executor is single threaded:
https://github.com/CorfuDB/CorfuDB/blob/003d656cfe17f35a3d6814bd9b31e1c16e679b6b/infrastructure/src/main/java/org/corfudb/infrastructure/LogReplicationServer.java#L77-L84
The gRPC implemented handlers are only used in tests, previously it used default thread pool executor, which could lead to ACK mismatch at sender side.

xcchang commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -427,8 +427,12 @@ public void onNext(CorfuStreamEntries results) {
+                    long lastConfigId = clusterManager.configId;
                     clusterManager.getClusterManagerCallback()
                             .applyNewTopologyConfig(clusterManager.generateConfigWithBackup());
+                    while (clusterManager.configId != lastConfigId + 1) {
```
Sorry I didn't follow the thread pool part, were you referring to the 2-thead worker pool for executing stream tasks? I thought by blocking the onNext method the new delta for this stream won't be scheduled for consuming (until currrent onNext exits)?

xcchang commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -1630,9 +1630,6 @@ public void testBackupRestoreWorkflow() throws Exception {
         updateTopology(activeCorfuStore, DefaultClusterManager.OP_BACKUP);
         log.info("Change the topology!!!");
 
-        TimeUnit.SECONDS.sleep(shortInterval);
```
yeah the code path is not interacted directly, and I was trying to achieve such workflow:
* Test method write to config table for OP_BACKUP
* Stream listener consumes update for OP_BACKUP, and invokes clusterManager's call back for applying new topology, and blocks until the new topology appears in DefaultClusterManager (DefaultClusterManager is registered in the stream listener)
* Test method write to config table for OP_ENFORCE_SNAPSHOT_FULL_SYNC (could happen before stream listener finishes consuming OP_BACKUP)
* Stream listener consumes update for OP_ENFORCE_SNAPSHOT_FULL_SYNC

Actually I think the second sleep (several lines below) could also be removed ... as we will block for the forced snapshot sync by checking table size

For the CountdownLatch method, if I understand correctly, it's also for blocking until topology change complete? but actually we didn't expose DefaultClusterManager to the test class, we only exposed some static fields like the OPs. So I think it will also work but could introduce more change...

shama358 replied:
i was thinking of having a countdown latch or similar construct as well instead of the while loop.
I think conditionVariable is very apt here.

Everytime its OP_BACKUP, wait on the conditionVariable, and once the configId in the defaultClusterManager is updated, notify the conditionVariable.

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Condition.html

xcchang replied:
Changed to Condition, also for other topology config change cases. Thanks!

souravgarg commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -434,6 +452,21 @@ public void onNext(CorfuStreamEntries results) {
+            if (clusterManager.configId != lastConfigId) {
+                try {
+                    clusterManager.topologyUpdateLock.lock();
+                    while (clusterManager.topologyConfig.getTopologyConfigID() != lastConfigId + 1) {
```
Just confirming if this loop for handling spurious wake in condition variable?

xcchang replied:
In my opinion the atomicity we want to achieve here is: when the stream listener consumes a topology change msg, it guarantees the topology in CorfuReplicationClusterManagerBaseAdapter is actually updated (before exit the onNext() method), but it won't block until the consumption of this event in discovery service, which I think also aligns with the real use case?

From the correctness perspective, I think we are fine here. As the topology in the adapter in the source of truth for new topology, and discovery service is doing some role-based setup according to it. Also, as mentioned in previous comment, our implementation of this DefaultClusterManager also fetches topology from the adapter.

souravgarg reviewed (commented):
Yes, condition variable seems much better here.

annym commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java:
```diff
@@ -369,6 +378,14 @@ public void run() {
                 try {
                     TopologyDescriptor newConfig = queue.take();
                     clusterManager.updateTopologyConfig(newConfig.convertToMessage());
```
+1

xcchang replied:
Actually I thought this point was the place we want to reach. After the returning of 
```
clusterManager.updateTopologyConfig(newConfig.convertToMessage());
```
we updated the topology cached in DefaultClusterManager (CorfuReplicationClusterManagerBaseAdapter), which is used to provide the topology for discovery service, and in this test case also used for following OP_ENFORCE_SNAPSHOT_FULL_SYNC
https://github.com/CorfuDB/CorfuDB/blob/dcab1d42738ba183bdd04a5feffd6a3b9dad4594/infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java#L418-L421

pankti-m commented on infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/transport/sample/GRPCLogReplicationServerChannelAdapter.java:
```diff
@@ -36,7 +37,8 @@ public GRPCLogReplicationServerChannelAdapter(ServerContext serverContext, LogRe
         this.service = new GRPCLogReplicationServerHandler(router);
         this.port = Integer.parseInt((String) serverContext.getServerConfig().get("<port>"));
-        this.server = ServerBuilder.forPort(port).addService(service).build();
+        this.server = ServerBuilder.forPort(port).addService(service)
```
I am fine with this change as it does look like the server will utilize this executorService.  Can you add a comment explaining that if multiple requests are received concurrently, ordering will not be guaranteed.  The default executorService uses a thread pool with multiple threads([GrpcUtils.class](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/GrpcUtil.java#L549)

pankti-m commented on test/src/test/java/org/corfudb/integration/CorfuReplicationClusterConfigIT.java:
```diff
@@ -1626,16 +1644,18 @@ public void testBackupRestoreWorkflow() throws Exception {
-
-        TimeUnit.SECONDS.sleep(shortInterval);
-
+        while (persistedTopologyConfigId >= getSinkPersistedTopologyConfigId()) {
```
Why do we check the topology config id on the Sink?  How will its topology config id change if OP_BACKUP was written and processed on the Source cluster?

xcchang committed:
Flaky test fix:

* CorfuReplicationClusterConfigIT.testBackupRestoreWorkflow: Block
  until new topology updated and persisted into LR metadata table
  at the backup cluster (new source)
